### PR TITLE
DM-20371: Create new shift/rot only WCS fitter

### PR DESCRIPTION
**Pull request by morriscb**
Initial try at a new, stripped down fitter for use with @parejkoj 's new distortion model code. Not meant for full review but for @parejkoj 's information if he would like to take up the ticket while I am on vacation.

### CODE REVIEW

File: tests/test_fitAffineWcsTask.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
+# LSST Data Management System
+# Copyright 2008, 2009, 2010 LSST Corporation.
```
Comment (TallJimbo) [inline tests/test_fitAffineWcsTask.py]: Should copy this copyright blurb from a newer file (like one of the new modules you added).

File: python/lsst/meas/astrom/fitAffineWcs.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        wcs_maker = TransformedSkyWcsMaker(initWcs)
+
+        ref_points = []
+        src_pixels = []
+        offset_dir = 0
+        offset_dist = 0
+        # Grab reference coordinates and source centroids. Compute the average
+        # direction and separation between the reference and the sources.
+        # I'm not sure if bearingTo should be computed from the src to ref
+        # or ref to source.
```
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/meas/astrom/fitAffineWcs.py]: It's resolved. The code is fine as is. There are a few comments round the code that were for @parejkoj to look into if he had time over the 2 weeks I was gone.

File: python/lsst/meas/astrom/fitAffineWcs.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        if offset_dir > 180:
+            offset_dir = offset_dir - 360
+        self.log.debug("Initial shift guess: Direction: %.3f, Dist %.3f..." %
+                       (offset_dir, offset_dist))
+
+        # Best performing fitter in scipy tried so far (vs. default settings in
+        # minimize). Fits all current test cases with a scatter of a most 0.15
+        # arcseconds. exits early because of the xTol value which cannot be
+        # disabled in scipy1.2.1. Matrix starting values are non-zero as this
+        # results in better fit off-diagonal terms.
```
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/meas/astrom/fitAffineWcs.py]: This was from the unittests copied from fitTanSip code which as 1 arcsecond pixels. I think things would be a bit better with the newer versions and removed tolerances as the code tends to exit quickly. With the unittests written in this ticket, the scatter is much much smaller and passes the test criteria without modification.

File: python/lsst/meas/astrom/fitAffineWcs.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+        # Grab the order of the frames by index.
+        domains = self.frame_dict.getAllDomains()
+        self.frame_idxs = np.sort([self.frame_dict.getIndex(domain)
+                                   for domain in domains])
+        self.frame_min = np.min(self.frame_idxs)
+        self.frame_max = np.max(self.frame_idxs)
+
+        # Find frame just before the final mapping to sky and store those
+        # indices and mappings for later.
```
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/meas/astrom/fitAffineWcs.py]: I mean that could be, however for now to use with the generic stack produced WCS, this is what I came up with to catch most cases. Most of the code in the stack seems to flatten the transform as `pixels->IWC->Sky`. If this changes and Parejko's new code will have more Frames, the code above will still work.

File: python/lsst/meas/astrom/fitAffineWcs.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        # Sky origin position. This is simply the final undistorted tangent
+        # plane to sky. The PIXELS to SKY map will be become our IWC to SKY
+        # map and gives us our final shift position.
+        iwcs_to_sky_wcs = makeSkyWcs(
+            Point2D(0., 0.),
+            self.origin.offset(crval_offset[0] * degrees,
+                               crval_offset[1] * arcseconds),
+            np.array([[1., 0.], [0., 1.]]))
+        iwc_to_sky_map = iwcs_to_sky_wcs.getFrameDict().getMapping("PIXELS",
+                                                                   "SKY")
```
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/meas/astrom/fitAffineWcs.py]: I think this requires a deeper change the C++ API of SkyWcs and the Sky frame. Sky frame is pretty special it seems and the code you see here is taken from unittests others wrote to create exactly this case. I believe the one I took here was written by @parejkoj and used in jointcal?

**Comment by morriscb** [comment]
Take a look at the changes. I've changed all the method and class variables to camel case. Is that enough?

I have one last improvement I thought of as the initial guess for the angle and distance for the fitter is computed naively and wrong. I'll have to do some testing tomorrow on this.

File: python/lsst/meas/astrom/fitAffineWcs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/meas/astrom/fitAffineWcs.py]: Please remove those comments if you haven't already.

File: python/lsst/meas/astrom/fitAffineWcs.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/meas/astrom/fitAffineWcs.py]: :+1:  Working with generic stack-produced WCSs is probably more important than named frames vs indices.  But I'd also like to get @parejkoj's thoughts on this, if only to wonder whether to bring this up again after this ticket is merged: will we have a named frame that can consistently be used for this purpose in all WCSs we produced (including, but not limited to the new initial ones)?

File: python/lsst/meas/astrom/fitAffineWcs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/meas/astrom/fitAffineWcs.py]: Yeah, I'm not advocating addressing this on this ticket.  But if we've got the same workaround appearing in jointcal as well, that'd make it even more important to factor it out into its own routine.

**Comment by TallJimbo** [comment]
> I've changed all the method and class variables to camel case. Is that enough?

Not up to me.  I don't think that as a reviewer I have the authority to grant exceptions to the style guide for no particular reason, especially in new code.

Sorry - I hate playing style police and I'm not a huge fan of camelCase either, and I personally very much agree that making the public interfaces conform is what matters most.  If you want to RFC allowing more snake_case, I think it'd probably get a lot of support (and depending on how we handle migrating existing code (or not) I'd probably get behind it, too).

File: python/lsst/meas/astrom/fitAffineWcs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (morriscb) [reply python/lsst/meas/astrom/fitAffineWcs.py]: Already is.

**Comment by morriscb** [comment]
Just finished up. Take another look if you want and I'll merge and close the ticket.

File: python/lsst/meas/astrom/fitAffineWcs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/meas/astrom/fitAffineWcs.py]: Russell suggested to me that the easiest way to create the correct IWC->SKY `Mapping` is to create a blank SkyWcs centered at the CRVAL you want and extract the Mapping and Frame from that. I don't know if we need another way to go about it or not: this is only two lines, and is a pretty specialized use.

[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/meas/astrom/fitAffineWcs.py]: This is the relevant code in jointcal:

https://github.com/lsst/jointcal/blob/master/src/ConstrainedAstrometryModel.cc#L234

File: python/lsst/meas/astrom/fitAffineWcs.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/meas/astrom/fitAffineWcs.py]: I dislike this approach, too. `makeSkyWcsFrameDict` (which is used under the hood to construct the new "initial" WCS) loses the names of the intermediate frames; we would normally want to put the new Transform right after `FIELD_ANGLE`, but that becomes `IWC` (which is the only "intermediate" frame that FITS WCS knows about). I filed [DM-20315](https://jira.lsstcorp.org/browse/DM-20315) about this issue.

At minimum, we should file a ticket triggered on DM-20315 to cleanup this code to insert the transformations between `FIELD_ANGLE` and `IWC` explicitly. That would make it incompatible with SkyWcs objects created in other ways, but I think that's fine: we're already well past the point of being "true FITS WCS" compatible with this work anyway.

I think we probably could make the names explicit here by explicitly inserting the new Transform between IWC and Sky, and correcting that to be between FIELD_ANGLE and IWC once 20315 is done.

**Pull request merged**